Terraform Crash

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to Terraform Cloud/Enterprise, please contact tf-cloud@hashicorp.support.
If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://registry.terraform.io/browse/providers.

To fix problems, we need clear reproduction cases - we need to be able to see it happen locally. A reproduction case is ideally something a Terraform Core engineer can git-clone or copy-paste and run immediately, without inventing any details or context. 

* A short example can be directly copy-pasteable; longer examples should be in separate git repositories, especially if multiple files are needed
* Please include all needed context. For example, if you figured out that an expression can cause a crash, put the expression in a variable definition or a resource
* Set defaults on (or omit) any variables. The person reproducing it should not need to invent variable settings
* If multiple steps are required, such as running terraform twice, consider scripting it in a simple shell script. Providing a script can be easier than explaining what changes to make to the config between runs.
* Omit any unneeded complexity: remove variables, conditional statements, functions, modules, providers, and resources that are not needed to trigger the bug
* When possible, use the [null resource](https://www.terraform.io/docs/providers/null/resource.html) provider rather than a real provider in order to minimize external dependencies. We know this isn't always feasible. The Terraform Core team doesn't have deep domain knowledge in every provider, or access to every cloud platform for reproduction cases.

-->

### Terraform Version
<!---
Run `terraform version` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
...
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For Terraform configs larger than a few resources, or that involve multiple files, please make a GitHub repository that we can clone, rather than copy-pasting multiple files in here. For security, you can also encrypt the files using our GPG public key at https://www.hashicorp.com/security.
-->

```terraform
...
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->

### Expected Behavior
<!--
What should have happened?
-->

### Actual Behavior
<!--
What actually happened?
-->

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->
[crash.769144250.log](https://github.com/hashicorp/terraform/files/6908039/crash.769144250.log)
2021-07-30T09:37:56.848-0400 [INFO]  Terraform version: 1.0.3
2021-07-30T09:37:56.849-0400 [INFO]  Go runtime version: go1.16.4
2021-07-30T09:37:56.849-0400 [INFO]  CLI args: []string{"C:\\Terraform\\terraform.exe", "plan"}
2021-07-30T09:37:56.851-0400 [TRACE] Stdout is a terminal of width 94
2021-07-30T09:37:56.851-0400 [TRACE] Stderr is a terminal of width 94
2021-07-30T09:37:56.851-0400 [TRACE] Stdin is a terminal
2021-07-30T09:37:56.854-0400 [DEBUG] Attempting to open CLI config file: C:\Users\User\AppData\Roaming\terraform.rc
2021-07-30T09:37:56.854-0400 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2021-07-30T09:37:56.855-0400 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2021-07-30T09:37:56.855-0400 [DEBUG] ignoring non-existing provider search directory C:\Users\User\AppData\Roaming\terraform.d\plugins
2021-07-30T09:37:56.856-0400 [DEBUG] ignoring non-existing provider search directory C:\Users\User\AppData\Roaming\HashiCorp\Terraform\plugins
2021-07-30T09:37:56.856-0400 [INFO]  CLI command args: []string{"plan"}

panic: runtime error: slice bounds out of range [:-2]

goroutine 1 [running]:
github.com/hashicorp/terraform/internal/command/format.appendSourceSnippets(0x15a8f638, 0x1561ae10, 0x1580c3a8)
	/home/circleci/project/project/internal/command/format/diagnostic.go:253 +0x8e1
github.com/hashicorp/terraform/internal/command/format.DiagnosticFromJSON(0x1561ae10, 0x1580c3a8, 0x5e, 0x1561ae10, 0x1586eea0)
	/home/circleci/project/project/internal/command/format/diagnostic.go:77 +0x1cc
github.com/hashicorp/terraform/internal/command/format.Diagnostic(0x2342f1c, 0x15617f50, 0x1586eea0, 0x1580c3a8, 0x5e, 0x1df, 0x0)
	/home/circleci/project/project/internal/command/format/diagnostic.go:30 +0x4d
github.com/hashicorp/terraform/internal/command/views.(*View).Diagnostics(0x15404a60, 0x15744d70, 0x2, 0x2)
	/home/circleci/project/project/internal/command/views/view.go:114 +0x2c9
github.com/hashicorp/terraform/internal/command/views.(*PlanHuman).Diagnostics(0x1580c6b8, 0x15744d70, 0x2, 0x2)
	/home/circleci/project/project/internal/command/views/plan.go:65 +0x3b
github.com/hashicorp/terraform/internal/command.(*PlanCommand).Run(0x15aaa4b0, 0x154940d8, 0x0, 0x0, 0x1580c438)
	/home/circleci/project/project/internal/command/plan.go:69 +0x770
github.com/mitchellh/cli.(*CLI).Run(0x156213f0, 0x1556ede0, 0x1f1c232, 0x4)
	/home/circleci/go/pkg/mod/github.com/mitchellh/cli@v1.1.2/cli.go:262 +0x336
main.wrappedMain(0x0)
	/home/circleci/project/project/main.go:384 +0x1014
main.realMain(0x0)
	/home/circleci/project/project/main.go:115 +0x475
main.main()
	/home/circleci/project/project/main.go:58 +0x1a

--- a/internal/command/format/diagnostic.go
+++ b/internal/command/format/diagnostic.go
@@ -250,6 +250,21 @@ func appendSourceSnippets(buf *bytes.Buffer, diag *viewsjson.Diagnostic, color *
 			}
 		}
 
+		// If either start or end is out of range for the code buffer then
+		// we'll cap them at the bounds just to avoid a panic, although
+		// this would happen only if there's a bug in the code generating
+		// the snippet objects.
+		if start < 0 {
+			start = 0
+		} else if start > len(code) {
+			start = len(code)
+		}
+		if end < 0 {
+			end = 0
+		} else if end > len(code) {
+			end = len(code)
+		}
+
 		before, highlight, after := code[0:start], code[start:end], code[end:]
 		code = fmt.Sprintf(color.Color("%s[underline]%s[reset]%s"), before, highlight, after)
 

--- a/internal/command/views/json/diagnostic.go
+++ b/internal/command/views/json/diagnostic.go
@@ -220,12 +220,23 @@ func NewDiagnostic(diag tfdiags.Diagnostic, sources map[string][]byte) *Diagnost
 			// to the code snippet string.
 			start := highlightRange.Start.Byte - codeStartByte
 			end := start + (highlightRange.End.Byte - highlightRange.Start.Byte)
-			if start > len(codeStr) {
+
+			// We can end up with some quirky results here in edge cases like
+			// when a source range starts or ends at a newline character,
+			// so we'll cap the results at the bounds of the highlight range
+			// so that consumers of this data don't need to contend with
+			// out-of-bounds errors themselves.
+			if start < 0 {
+				start = 0
+			} else if start > len(codeStr) {
 				start = len(codeStr)
 			}
-			if end > len(codeStr) {
+			if end < 0 {
+				end = 0
+			} else if end > len(codeStr) {
 				end = len(codeStr)
 			}
+
 			diagnostic.Snippet.HighlightStartOffset = start
 			diagnostic.Snippet.HighlightEndOffset = end
 

--- a/internal/command/views/json/diagnostic_test.go
+++ b/internal/command/views/json/diagnostic_test.go
@@ -28,7 +28,8 @@ func TestNewDiagnostic(t *testing.T) {
   }
 }
 `),
-		"short.tf": []byte("bad source code"),
+		"short.tf":       []byte("bad source code"),
+		"odd-comment.tf": []byte("foo\n\n#\n"),
 		"values.tf": []byte(`[
   var.a,
   var.b,
@@ -281,6 +282,51 @@ func TestNewDiagnostic(t *testing.T) {
 					HighlightStartOffset: 15,
 					HighlightEndOffset:   16,
 					Values:               []DiagnosticExpressionValue{},
+				},
+			},
+		},
+		"error whose range starts at a newline": {
+			&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid newline",
+				Detail:   "How awkward!",
+				Subject: &hcl.Range{
+					Filename: "odd-comment.tf",
+					Start:    hcl.Pos{Line: 2, Column: 5, Byte: 4},
+					End:      hcl.Pos{Line: 3, Column: 1, Byte: 6},
+				},
+			},
+			&Diagnostic{
+				Severity: "error",
+				Summary:  "Invalid newline",
+				Detail:   "How awkward!",
+				Range: &DiagnosticRange{
+					Filename: "odd-comment.tf",
+					Start: Pos{
+						Line:   2,
+						Column: 5,
+						Byte:   4,
+					},
+					End: Pos{
+						Line:   3,
+						Column: 1,
+						Byte:   6,
+					},
+				},
+				Snippet: &DiagnosticSnippet{
+					Code:      `#`,
+					StartLine: 2,
+					Values:    []DiagnosticExpressionValue{},
+
+					// Due to the range starting at a newline on a blank
+					// line, we end up stripping off the initial newline
+					// to produce only a one-line snippet. That would
+					// therefore cause the start offset to naturally be
+					// -1, just before the Code we returned, but then we
+					// force it to zero so that the result will still be
+					// in range for a byte-oriented slice of Code.
+					HighlightStartOffset: 0,
+					HighlightEndOffset:   1,
 				},
 			},
 		},

--- a/internal/command/views/json/testdata/diagnostic/error-whose-range-starts-at-a-newline.json
+++ b/internal/command/views/json/testdata/diagnostic/error-whose-range-starts-at-a-newline.json
@@ -1,0 +1,26 @@
+{
+  "severity": "error",
+  "summary": "Invalid newline",
+  "detail": "How awkward!",
+  "range": {
+    "filename": "odd-comment.tf",
+    "start": {
+      "line": 2,
+      "column": 5,
+      "byte": 4
+    },
+    "end": {
+      "line": 3,
+      "column": 1,
+      "byte": 6
+    }
+  },
+  "snippet": {
+    "context": null,
+    "code": "#",
+    "start_line": 2,
+    "highlight_start_offset": 0,
+    "highlight_end_offset": 1,
+    "values": []
+  }
+}

